File: engine/crates/blocus-python/python_tests/test_board_and_piece_api.py
import json
import uuid

import pytest

from blocus_engine import (
    BlocusEngine,
    EngineError,
    GameConfig,
    GameMode,
    GameState,
    InputError,
    PlaceCommand,
    PlayerColor,
    PlayerSlots,
    ScoringMode,
)


BLUE_PLAYER = "00000000-0000-0000-0000-00000000000a"
YELLOW_GREEN_PLAYER = "00000000-0000-0000-0000-000000000014"
GAME_ID = "00000000-0000-0000-0000-000000000001"


def make_config():
    return GameConfig(
        game_id=GAME_ID,
        mode=GameMode.TWO_PLAYER,
        scoring=ScoringMode.BASIC,
        turn_order=[
            PlayerColor.BLUE,
            PlayerColor.YELLOW,
            PlayerColor.RED,
            PlayerColor.GREEN,
        ],
        player_slots=PlayerSlots.two_player(
            BLUE_PLAYER,
            YELLOW_GREEN_PLAYER,
        ),
    )


def make_place_command(state):
    return PlaceCommand(
        command_id=str(uuid.uuid4()),
        game_id=state.game_id,
        player_id=BLUE_PLAYER,
        color=PlayerColor.BLUE,
        piece_id=0,
        orientation_id=0,
        row=0,
        col=0,
    )


def make_state():
    return BlocusEngine().initialize_game(make_config())


def test_game_state_exposes_empty_board_views():
    state = make_state()

    assert state.board_is_empty is True
    assert state.cell(0, 0) is None
    assert state.occupied_cells() == []
    assert state.occupied_cells(PlayerColor.BLUE) == []

    matrix = state.board_matrix()
    assert len(matrix) == 20
    assert all(len(row) == 20 for row in matrix)
    assert all(cell is None for row in matrix for cell in row)

    counts = dict(state.board_counts())
    assert counts[PlayerColor.BLUE] == 0
    assert counts[PlayerColor.YELLOW] == 0
    assert counts[PlayerColor.RED] == 0
    assert counts[PlayerColor.GREEN] == 0


def test_game_state_exposes_board_after_place_move():
    engine = BlocusEngine()
    state = engine.initialize_game(make_config())

    result = engine.apply(state, make_place_command(state))
    next_state = result.next_state

    assert next_state.board_is_empty is False
    assert next_state.cell(0, 0) == PlayerColor.BLUE
    assert next_state.cell(0, 1) is None

    blue_cells = next_state.occupied_cells(PlayerColor.BLUE)
    assert len(blue_cells) == 1
    assert blue_cells[0].row == 0
    assert blue_cells[0].col == 0
    assert blue_cells[0].color == PlayerColor.BLUE

    all_cells = next_state.occupied_cells()
    assert all_cells == blue_cells

    counts = dict(next_state.board_counts())
    assert counts[PlayerColor.BLUE] == 1
    assert counts[PlayerColor.YELLOW] == 0


def test_cell_rejects_invalid_board_coordinates():
    state = make_state()

    with pytest.raises(InputError):
        state.cell(20, 0)

    with pytest.raises(InputError):
        state.cell(0, 20)


def test_engine_exposes_piece_repository_and_orientations():
    engine = BlocusEngine()

    pieces = engine.pieces()
    assert len(pieces) == 21

    first = engine.piece(0)
    assert first.id == 0
    assert first.name
    assert first.square_count >= 1
    assert first.orientation_count >= 1
    assert len(first.orientations) == first.orientation_count

    orientation = first.orientation(0)
    assert orientation.id == 0
    assert orientation.width >= 1
    assert orientation.height >= 1
    assert orientation.square_count == first.square_count
    assert len(orientation.cells) == first.square_count

    absolute_cells = orientation.cells_at(0, 0)
    assert absolute_cells == orientation.cells


def test_piece_lookup_rejects_invalid_ids():
    engine = BlocusEngine()
    piece = engine.piece(0)

    with pytest.raises(InputError):
        engine.piece(255)

    with pytest.raises(InputError):
        piece.orientation(255)


def test_orientation_cells_at_rejects_out_of_bounds_anchor():
    engine = BlocusEngine()
    orientation = engine.piece(0).orientation(0)

    with pytest.raises(InputError):
        orientation.cells_at(20, 0)

    with pytest.raises(InputError):
        orientation.cells_at(0, 20)


def test_game_state_exposes_inventory_views():
    state = make_state()

    assert state.used_piece_ids(PlayerColor.BLUE) == []
    assert state.available_piece_ids(PlayerColor.BLUE) == list(range(21))

    used_pieces = state.used_pieces(PlayerColor.BLUE)
    available_pieces = state.available_pieces(PlayerColor.BLUE)

    assert used_pieces == []
    assert len(available_pieces) == 21
    assert [piece.id for piece in available_pieces] == list(range(21))

    summary = state.inventory_summary(PlayerColor.BLUE)
    assert summary.color == PlayerColor.BLUE
    assert summary.used_piece_ids == []
    assert summary.available_piece_ids == list(range(21))
    assert summary.used_count == 0
    assert summary.available_count == 21
    assert summary.used_square_count == 0
    assert summary.remaining_square_count > 0
    assert summary.is_complete is False
    assert state.remaining_square_count(PlayerColor.BLUE) == summary.remaining_square_count


def test_inventory_views_update_after_move():
    engine = BlocusEngine()
    state = engine.initialize_game(make_config())

    result = engine.apply(state, make_place_command(state))
    next_state = result.next_state

    assert next_state.used_piece_ids(PlayerColor.BLUE) == [0]
    assert 0 not in next_state.available_piece_ids(PlayerColor.BLUE)

    used_pieces = next_state.used_pieces(PlayerColor.BLUE)
    assert [piece.id for piece in used_pieces] == [0]

    summary = next_state.inventory_summary(PlayerColor.BLUE)
    assert summary.used_piece_ids == [0]
    assert summary.used_count == 1
    assert summary.available_count == 20
    assert summary.used_square_count == 1
    assert summary.remaining_square_count == state.remaining_square_count(PlayerColor.BLUE) - 1


def test_engine_exposes_valid_moves_and_piece_filtered_valid_moves():
    engine = BlocusEngine()
    state = engine.initialize_game(make_config())

    moves = engine.get_valid_moves(state, BLUE_PLAYER, PlayerColor.BLUE)
    piece_moves = engine.get_valid_moves_for_piece(
        state,
        BLUE_PLAYER,
        PlayerColor.BLUE,
        0,
    )

    assert moves
    assert piece_moves
    assert all(move.piece_id == 0 for move in piece_moves)
    assert len(piece_moves) <= len(moves)

    assert engine.has_any_valid_move(state, BLUE_PLAYER, PlayerColor.BLUE) is True
    assert engine.has_any_valid_move_for_piece(
        state,
        BLUE_PLAYER,
        PlayerColor.BLUE,
        0,
    ) is True


def test_valid_moves_are_empty_for_wrong_player_and_finished_state():
    engine = BlocusEngine()
    state = engine.initialize_game(make_config())

    assert engine.get_valid_moves(state, YELLOW_GREEN_PLAYER, PlayerColor.BLUE) == []
    assert engine.has_any_valid_move(state, YELLOW_GREEN_PLAYER, PlayerColor.BLUE) is False

    data = json.loads(state.to_json())
    data["status"] = "finished"
    finished = GameState.from_json(json.dumps(data))

    assert engine.get_valid_moves(finished, BLUE_PLAYER, PlayerColor.BLUE) == []
    assert engine.has_any_valid_move(finished, BLUE_PLAYER, PlayerColor.BLUE) is False


def test_json_parser_rejects_invalid_status_instead_of_defaulting_to_in_progress():
    state = make_state()
    data = json.loads(state.to_json())
    data["status"] = "bogus"

    with pytest.raises(InputError):
        GameState.from_json(json.dumps(data))


def test_from_json_recomputes_hash_and_ignores_input_hash():
    state = make_state()
    data = json.loads(state.to_json())

    original_hash = data["hash"]
    data["hash"] = original_hash + 123456789

    restored = GameState.from_json(json.dumps(data))

    assert restored.hash == original_hash
    assert restored.hash != data["hash"]


def test_from_json_rejects_invalid_board_padding_bits():
    state = make_state()
    data = json.loads(state.to_json())

    lanes = data["board"]["blue"]["lanes"]
    lanes[0] = str(1 << 20)

    with pytest.raises(EngineError):
        GameState.from_json(json.dumps(data))


def test_from_json_rejects_overlapping_color_masks():
    state = make_state()
    data = json.loads(state.to_json())

    data["board"]["blue"]["lanes"][0] = "1"
    data["board"]["yellow"]["lanes"][0] = "1"

    with pytest.raises(EngineError):
        GameState.from_json(json.dumps(data))

File: engine/crates/blocus-core/tests/test_board_and_piece_api.rs
use uuid::Uuid;

use blocus_core::{
    BlocusEngine, BoardIndex, Command, CommandId, GameConfig, GameId, GameMode, GameStatus,
    OrientationId, PassCommand, PieceId, PlaceCommand, PlayerColor, PlayerId, PlayerSlots,
    ScoringMode, TurnOrder,
};

fn uuid(value: u128) -> Uuid {
    Uuid::from_u128(value)
}

fn player(value: u128) -> PlayerId {
    PlayerId::from_uuid(uuid(value))
}

fn game_id(value: u128) -> GameId {
    GameId::from_uuid(uuid(value))
}

fn command_id(value: u128) -> CommandId {
    CommandId::from_uuid(uuid(value))
}

fn piece_id(value: u8) -> PieceId {
    PieceId::try_new(value).unwrap_or_else(|_| panic!("piece id {value} should be valid"))
}

fn orientation_id(value: u8) -> OrientationId {
    OrientationId::try_new(value)
        .unwrap_or_else(|_| panic!("orientation id {value} should be valid"))
}

fn index(row: u8, col: u8) -> BoardIndex {
    BoardIndex::from_row_col(row, col)
        .unwrap_or_else(|_| panic!("row {row}, col {col} should be valid"))
}

fn two_player_slots() -> PlayerSlots {
    PlayerSlots::two_player(player(10), player(20))
        .unwrap_or_else(|_| panic!("two-player slots should be valid"))
}

fn two_player_config() -> GameConfig {
    GameConfig::try_new(
        game_id(1),
        GameMode::TwoPlayer,
        ScoringMode::Basic,
        TurnOrder::OFFICIAL_FIXED,
        two_player_slots(),
    )
    .unwrap_or_else(|_| panic!("two-player config should be valid"))
}

fn opening_place_command(
    state_game_id: GameId,
    command_value: u128,
    player_id: PlayerId,
    color: PlayerColor,
    piece_id: PieceId,
    orientation_id: OrientationId,
    anchor: BoardIndex,
) -> Command {
    Command::Place(PlaceCommand {
        command_id: command_id(command_value),
        game_id: state_game_id,
        player_id,
        color,
        piece_id,
        orientation_id,
        anchor,
    })
}

fn pass_command(
    state_game_id: GameId,
    command_value: u128,
    player_id: PlayerId,
    color: PlayerColor,
) -> Command {
    Command::Pass(PassCommand {
        command_id: command_id(command_value),
        game_id: state_game_id,
        player_id,
        color,
    })
}

#[test]
fn board_state_exposes_color_at_counts_and_occupied_cells() {
    let engine = BlocusEngine::new();
    let state = engine.initialize_game(two_player_config());

    assert!(state.board.is_empty());
    assert_eq!(state.board.occupied_count(PlayerColor::Blue), 0);
    assert_eq!(state.board.occupied_cells(PlayerColor::Blue), Vec::new());
    assert_eq!(state.board.occupied_cells_all(), Vec::new());

    let result = engine
        .apply(
            &state,
            opening_place_command(
                state.game_id,
                1,
                player(10),
                PlayerColor::Blue,
                piece_id(0),
                orientation_id(0),
                index(0, 0),
            ),
        )
        .unwrap_or_else(|error| panic!("first blue move should be legal: {error}"));

    let next = result.next_state;
    let corner = index(0, 0);

    assert!(!next.board.is_empty());
    assert_eq!(next.board.color_at(corner), Some(PlayerColor::Blue));
    assert_eq!(next.board.occupied_count(PlayerColor::Blue), 1);
    assert_eq!(next.board.occupied_count(PlayerColor::Yellow), 0);

    assert_eq!(next.board.occupied_cells(PlayerColor::Blue), vec![corner]);
    assert_eq!(
        next.board.occupied_cells_all(),
        vec![(PlayerColor::Blue, corner)]
    );
}

#[test]
fn engine_exposes_piece_repository_metadata() {
    let engine = BlocusEngine::new();

    let pieces = engine.pieces();
    assert_eq!(pieces.len(), usize::from(blocus_core::PIECE_COUNT));

    let monomino = engine.piece(piece_id(0));

    assert_eq!(monomino.id(), piece_id(0));
    assert_eq!(monomino.name(), "I1");
    assert_eq!(monomino.square_count(), 1);
    assert_eq!(monomino.orientation_count(), 1);

    let orientation = monomino
        .orientation(orientation_id(0))
        .unwrap_or_else(|| panic!("monomino orientation 0 should exist"));

    assert_eq!(orientation.id(), orientation_id(0));
    assert_eq!(orientation.shape().square_count(), 1);
    assert_eq!(orientation.shape().cells(), vec![(0, 0)]);
}

#[test]
fn engine_exposes_all_valid_moves_and_piece_filtered_valid_moves() {
    let engine = BlocusEngine::new();
    let state = engine.initialize_game(two_player_config());

    let blue_player = player(10);
    let monomino = piece_id(0);

    let all_moves = engine
        .get_valid_moves(&state, blue_player, PlayerColor::Blue)
        .unwrap_or_else(|error| panic!("valid move generation should succeed: {error}"));

    let monomino_moves = engine
        .get_valid_moves_for_piece(&state, blue_player, PlayerColor::Blue, monomino)
        .unwrap_or_else(|error| panic!("piece-filtered move generation should succeed: {error}"));

    assert!(!all_moves.is_empty());
    assert!(!monomino_moves.is_empty());
    assert!(
        monomino_moves
            .iter()
            .all(|legal_move| legal_move.piece_id == monomino)
    );
    assert!(monomino_moves.len() <= all_moves.len());

    assert_eq!(
        engine.has_any_valid_move(&state, blue_player, PlayerColor::Blue),
        Ok(true)
    );

    assert_eq!(
        engine.has_any_valid_move_for_piece(&state, blue_player, PlayerColor::Blue, monomino),
        Ok(true)
    );
}

#[test]
fn valid_moves_are_empty_for_wrong_player_or_finished_game() {
    let engine = BlocusEngine::new();
    let state = engine.initialize_game(two_player_config());

    let wrong_player_moves = engine
        .get_valid_moves(&state, player(20), PlayerColor::Blue)
        .unwrap_or_else(|error| panic!("wrong controller should not corrupt movegen: {error}"));

    assert!(wrong_player_moves.is_empty());

    let mut finished = state.clone();
    finished.status = GameStatus::Finished;

    let finished_moves = engine
        .get_valid_moves(&finished, player(10), PlayerColor::Blue)
        .unwrap_or_else(|error| panic!("finished game should not corrupt movegen: {error}"));

    assert!(finished_moves.is_empty());
}

#[test]
fn piece_inventory_tracks_used_and_available_pieces_after_move() {
    let engine = BlocusEngine::new();
    let state = engine.initialize_game(two_player_config());

    let monomino = piece_id(0);

    assert!(!state.inventories[PlayerColor::Blue.index()].is_used(monomino));
    assert_eq!(
        state.used_piece_ids(PlayerColor::Blue),
        Vec::<PieceId>::new()
    );
    assert_eq!(state.available_piece_ids(PlayerColor::Blue).len(), 21);

    let result = engine
        .apply(
            &state,
            opening_place_command(
                state.game_id,
                2,
                player(10),
                PlayerColor::Blue,
                monomino,
                orientation_id(0),
                index(0, 0),
            ),
        )
        .unwrap_or_else(|error| panic!("first blue move should be legal: {error}"));

    let next = result.next_state;

    assert!(next.inventories[PlayerColor::Blue.index()].is_used(monomino));
    assert_eq!(next.inventories[PlayerColor::Blue.index()].used_count(), 1);
    assert_eq!(
        next.inventories[PlayerColor::Blue.index()].available_count(),
        u32::from(blocus_core::PIECE_COUNT) - 1
    );

    assert_eq!(next.used_piece_ids(PlayerColor::Blue), vec![monomino]);
    assert!(
        !next
            .available_piece_ids(PlayerColor::Blue)
            .contains(&monomino)
    );

    let yellow_monomino_moves = engine
        .get_valid_moves_for_piece(&next, player(20), PlayerColor::Yellow, monomino)
        .unwrap_or_else(|error| panic!("yellow monomino move query should succeed: {error}"));

    assert!(!yellow_monomino_moves.is_empty());

    let used_blue_monomino_moves = engine
        .get_valid_moves_for_piece(&next, player(10), PlayerColor::Blue, monomino)
        .unwrap_or_else(|error| panic!("used piece move query should succeed: {error}"));

    assert!(used_blue_monomino_moves.is_empty());
}

#[test]
fn pass_probe_uses_valid_move_generation() {
    let engine = BlocusEngine::new();
    let state = engine.initialize_game(two_player_config());

    let result = engine.apply(
        &state,
        pass_command(state.game_id, 3, player(10), PlayerColor::Blue),
    );

    assert!(
        result.is_err(),
        "passing must be rejected while a legal move exists"
    );
}
